### PR TITLE
Integration

### PR DESCRIPTION
**Pull request by sblackler**
This PR represents the work done by @smudge202 & myself to clean up the Memo code ready for integration into Compose.Remoting

Having discussed this with @smudge202, there are a few issues outstanding which I will raise as a summary under the Compose.Remoting repo.

### CODE REVIEW

**Comment by sblackler** [comment]
App Veyor build fails due to the following error message: 

Unknown test framework: Could not find xunit.dll or xunit.execution.desktop.dll

**Comment by smudge202** [comment]
The sblackler/compose-remoting#72 PR will resolve the build issue (want to accept that on your side @sblackler ?)

**Comment by sblackler** [comment]
Done

File: src/Compose.Remoting/Providers/DefaultTypeNamingProvider.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace Compose.Remoting.Providers
+{
+    public sealed class DefaultTypeNamingProvider : ITypeNamingProvider
+    {
+        public string GetTypeName(object entity)
+        {
+            return entity.GetType().AssemblyQualifiedName;
```
Comment (mattridgway) [inline src/Compose.Remoting/Providers/DefaultTypeNamingProvider.cs]: Exception could get thrown here if entity was null

File: src/Compose.Remoting/Exceptions/MessageHandlerNotRegisteredException.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+{
+    [Serializable]
+    public class MessageHandlerNotRegisteredException : Exception
+    {
+        public MessageHandlerNotRegisteredException()
+        {
+        }
+
+        public MessageHandlerNotRegisteredException(Type messageType)
+            : base($"The message handler for type '{messageType.Name}', could not be found")
```
Comment (mattridgway) [inline src/Compose.Remoting/Exceptions/MessageHandlerNotRegisteredException.cs]: possible exception here is messagetype is null

File: src/Compose.Remoting/Exceptions/InvalidMessageException.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Compose.Remoting.Exceptions
+{
+    [Serializable]
+    public class InvalidMessageException : Exception
+    {
+        public InvalidMessageException(Message message) : base("The message supplied is invalid in its current state")
```
Comment (mattridgway) [inline src/Compose.Remoting/Exceptions/InvalidMessageException.cs]: you are not using the Message parameter

Comment (sblackler) [inline src/Compose.Remoting/Exceptions/InvalidMessageException.cs]: See https://github.com/4Com/compose-remoting/issues/11 - Going to add it back later.

File: src/Compose.Remoting/Exceptions/MessageHandlerNotRegisteredException.cs (hunk 2 above)
Comment (sblackler) [inline src/Compose.Remoting/Exceptions/MessageHandlerNotRegisteredException.cs]: IIRC, base will be called before our constructor?

File: src/Compose.Remoting/Providers/DefaultTypeNamingProvider.cs (hunk 1 above)
Comment (sblackler) [inline src/Compose.Remoting/Providers/DefaultTypeNamingProvider.cs]: Fixed in: https://github.com/sblackler/compose-remoting/commit/2b06d4152ed5cc70f7a9077efbbaed3cf0b156e8

File: src/Compose.Remoting/MessageHandling/DefaultMessageGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+			_compressionProvider = compressionProvider;
+			_serializationProvider = serializationProvider;
+			_encryptionProvider = encryptionProvider;
+			_typeNamingProvider = typeNamingProvider;
+		}
+
+		public OutboundMessage GenerateMessage<T>(T entity)
+		{
+			var typeName = _typeNamingProvider.GetTypeName(entity);
```
Comment (mattridgway) [inline src/Compose.Remoting/MessageHandling/DefaultMessageGenerator.cs]: might seem small - but why not put this below, inline, when you are creating the object? You set all the other properties this way.

File: src/Compose.Remoting/Exceptions/InvalidMessageException.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Compose.Remoting.Exceptions
```
Comment (mattridgway) [inline src/Compose.Remoting/Exceptions/InvalidMessageException.cs]: I would move this to the Compose.Remoting namespace, and remove from Exceptions folder. Exceptions should live with the classes where they are used

File: src/Compose.Remoting/Exceptions/InvalidMessageException.cs (hunk 3 above)
Comment (mattridgway) [inline src/Compose.Remoting/Exceptions/InvalidMessageException.cs]: Good point, forgot that one

File: src/Compose.Remoting/Exceptions/MessageHandlerNotRegisteredException.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Compose.Remoting.Exceptions
```
Comment (mattridgway) [inline src/Compose.Remoting/Exceptions/MessageHandlerNotRegisteredException.cs]: I would move to Compose.Remoting.MessageHandling namespace and out of the exceptions folder

File: src/Compose.Remoting/Exceptions/ProviderNotRegisteredException.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Compose.Remoting.Exceptions
```
Comment (mattridgway) [inline src/Compose.Remoting/Exceptions/ProviderNotRegisteredException.cs]: I would move to the Compose.Remoting.Providers namespace and out of the exceptions folder

**Comment by mattridgway** [comment]
Could you update the ReadMe, to explain what the steps are in the pipline?

**Comment by sblackler** [comment]
Think I've covered it all, if i've missed something give me a shout

File: src/Compose.Remoting/MessageHandling/DefaultMessageProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                if (inboundInterceptor.Intercept(incomingMessage) == InterceptorResult.Block)
+                {
+                    return;
+                }
+            }
+
+            var handler = _messageHandlerFactory.Create(_typeNamingProvider.GetType(incomingMessage.TypeName));
+            handler.Handle(incomingMessage.Entity);
+
+            foreach (var outboundInterceptor in _outboundInterceptors)
```
Comment (mattridgway) [inline src/Compose.Remoting/MessageHandling/DefaultMessageProcessor.cs]: should the outbound interceptors be able to block, just like the incoming ones?

Comment (sblackler) [inline src/Compose.Remoting/MessageHandling/DefaultMessageProcessor.cs]: The purpose of the block is to stop passing to either the transport (on the outbound) and the message handler on the inbound. So i'd probably say not.

**Comment by mattridgway** [comment]
@sblackler / @smudge202 - happy to merge this into dev for now, I would really like to see a sample app (and the supporting classes needed for this) and test coverage next, before we merged this into master. Thoughts?

**Comment by sblackler** [comment]
I think the test coverage needs to be done before the merge to master. I think the Compose Api is next on the list (eg: .AsRemoting()). Then probably the sample app/test coverage. Finally some supporting documentation on the wiki before the merge to master. Whether building the sample app will help with the api I don't know. I'll defer to both of your judgements :)

**Comment by smudge202** [comment]
If you're happy for me to do the Compose API work, my intention would be to create the sample application. I prefer to _use_ the API in a sample app to work out the best/simplest/most user friendly way the API can be, then attempt to put the code in place to make the sample API work.

**Comment by sblackler** [comment]
Yeah go for it. Id like to see somewhere where I can do the following:

.RegisterMessageHandler<MessageType>(Action<MessageType> action)

And

.RegisterMessageHandler<TConcreteHandler, MessageType>()

This will allow for true unobtrusiveness with regards handlers as that's the only bit currently that requires an interface. We can use some dynamic generation to create what we need off of this. 

Probably best to copy this into an issue later

**Comment by smudge202** [comment]
Blank issue added to continue API discussion. Spam away.

https://github.com/4Com/compose-remoting/issues/14

**Pull request merged**